fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java

```diff
@@ -107,40 +104,61 @@ private static long fileLockValueToTimestamp(byte[] value) {
+                agilityContext.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_FILE_LOCK_SET,
+                        agilityContext.apply(aContext -> fileLockSet(isHeartbeat, aContext)
+                        ));
+            }
         }
 
-        public long fileLockGet() {
-            // return time stamp if exists, 0 if not
-            byte[] val = agilityContext.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_FILE_LOCK_GET, agilityContext.get(fileLockKey));
-            return fileLockValueToTimestamp(val);
+        private CompletableFuture<Void> fileLockSet(boolean isHeartbeat, FDBRecordContext aContext) {
+            final long nowMillis = System.currentTimeMillis();
+            return aContext.ensureActive().get(fileLockKey)
+                    .thenAccept(val -> {
+                        long existingTimeStamp = fileLockValueToTimestamp(val);
+                        if (isHeartbeat) {
+                            fileLockCheckHeartBeat(existingTimeStamp);
+                        } else {
+                            fileLockCheckNewLock(existingTimeStamp, nowMillis);
+                        }
+                        this.timeStampMillis = nowMillis;
+                        byte[] value = fileLockValue(timeStampMillis);
+                        aContext.ensureActive().set(fileLockKey, value);
+                    });
         }
 
-        public void fileLockClear(long timeStampMillis) {
+        private void fileLockCheckHeartBeat(long existingTimeStamp) {
+            if (existingTimeStamp == 0) {
+                throw new AlreadyClosedException("Lock file was deleted. This=" + this);
+            }
+            if (existingTimeStamp != timeStampMillis) {
+                throw new AlreadyClosedException("Lock file changed by an external force at " + existingTimeStamp + ". This=" + this);
+            }
+        }
+
+        private void fileLockCheckNewLock(long existingTimeStamp, long nowMillis) {
+            if (existingTimeStamp <= 0) {
+                // all clear
+                return;
+            }
+            if (existingTimeStamp > (nowMillis - timeWindowMilliseconds) &&
+                    existingTimeStamp < (nowMillis + timeWindowMilliseconds)) {
+                // Here: this lock is valid
+                throw new RecordCoreException("FileLock: Set: found old lock")
+                        .addLogInfo(LuceneLogMessageKeys.LOCK_EXISTING_TIMESTAMP, existingTimeStamp,
+                                LuceneLogMessageKeys.LOCK_DIRECTORY, this);
+            }
+            // Here: this lock is either too old, or in the future. Steal it
+            if (LOGGER.isWarnEnabled()) {
+                LOGGER.warn(KeyValueLogMessage.of("FileLock: Set: found old lock, discard it",
+                        LuceneLogMessageKeys.LOCK_EXISTING_TIMESTAMP, existingTimeStamp,
+                        LuceneLogMessageKeys.LOCK_DIRECTORY, this));
+            }
+        }
+
+
+        private void fileLockClear() {
```

Comment:
I just realized this doesn't work in the case where there is a heartbeat. Namely, it's possible that it does a heartbeat, and fails to commit. If it tries to clear the lock in a separate transaction, it won't know what the timestamp should be, since it already updated `this.timeStampMillis`. Switching it to use a uuid would allow `fileLockClear` to just check the `uuid`, and not the timestamp.

Reply:
Implemented: 
* Moved locks (heartbeat and clear) to be within the get. 
* Use UUID to identify self lock.